### PR TITLE
scripts: add an automatic range generator.

### PR DESCRIPTION
**Pull request by knz**
cc @jseldess @petermattis.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/11203)

### CODE REVIEW

**Comment by knz** [comment]
FWIW I'm getting 88 splits per second on this gce instance.

**Comment by petermattis** [comment]
Was there a specific use case for this? Curious what the motivation was. We have other tools which can generate large numbers of ranges (i.e. `allocsim` and `zerosum`), though none that do it in quite this way.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[scripts/gen-ranges.py, line 2 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11203#-KXHHuwYtSrCZEokmYxR:-KXHHuwYtSrCZEokmYxS:bnstu9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c35149ee84f3a3889aed1c99e00a1a032d1f9be/scripts/gen-ranges.py#L2)):*
> ```Python
> # Import the driver.
> import psycopg2
> ```

Why python? I suppose this might be tied into the motivation. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11203#-:-KXHHciCkkEyUSDPGanG:b6zxq68)*

**Comment by knz** [comment]
Use case: @jseldess was asking for this, and I wanted to quantify some of the missing variables in https://www.cockroachlabs.com/blog/memory-usage-cockroachdb/

Why python: why not? I'm not yet as fluent in go as in python, and this is not a product feature.

**Comment by jseldess** [comment]
This is great, @knz. But for the scalability demo, what we need is actually a bit different. After drastically reducing the max range size, we need a load generator that lets us fill up ranges and observe splitting (via the replica counts in the admin ui). Do you think one of the other tools @petermattis mentioned would better meet this use case?

**Comment by knz** [comment]
Yes, and so will our trusted `block_writer` utility too.

**Comment by knz** [comment]
Note that this range generator of mine will also let you observe range splitting.

**Comment by petermattis** [comment]
When I want to generate a lot of ranges I do something like:

```
~ roachdemo -n 3
...
~ echo -e "range_min_bytes: 1\nrange_max_bytes: $[256 * 1024]" | ./cockroach zone set .default --file=-
...
~ block_writer
```

This will set the range target size to 256 KB and `block_writer` will fill those ranges and force them to split very quickly.

**Comment by petermattis** [comment]
> Why python: why not? I'm not yet as fluent in go as in python, and this is not a product feature.

True, this isn't a product feature, but we've moved towards using Go over shell in various areas because of the better knowledge everyone has of Go. I'm not particularly opposed to using Python, though, just curious as to the motivation.

**Comment by bdarnell** [comment]
For the scalability demo, I don't think we want to tell people to set up a python environment and install psycopg2, etc. Maybe build something like block_writer in to `cockroach gen`? 

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11203#-:-KXPuw8aMkrSqJIyPTcE:b-ucnnnw)*

File: scripts/gen-ranges.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        sys.stdout.write(m)
+
+startt = time.time()
+nrows = 0
+nsplits = 0
+print "First value: %d" % first
+for i in xrange(first, first+count):
+        o('.')
+        try:
+                onestmt(conn, "insert into %s(x) values (%d)" % (tname, i))
```
Comment (bdarnell) [inline scripts/gen-ranges.py]: Also, is the insert necessary? I think it should be possible to split the table into zero-width ranges.

**Comment by jseldess** [comment]
@bdarnell, if we could build this into `cockroach gen`, that would be the best in terms of usability (no dependencies). But in the meantime, I was planning to feature the `block_writer` in Go, with a note for users who don't have Go installed that they can use [this SQL](https://github.com/cockroachdb/cockroach/issues/11208) that Raphael pulled together. Thoughts?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11203#-:-KXS1D0sU9FSDGbg4hoG:b-6p8aa7)*

**Comment by knz** [comment]
The trick I proposed is not possible any more since Ben merged  #15350.
I am closing this issue since we made good progress on load generators.

**Pull request closed without merging**